src/video/wayland/SDL_waylandwindow.c

```diff
@@ -448,8 +491,109 @@ static const struct xdg_toplevel_listener toplevel_listener_xdg = {
     handle_close_xdg_toplevel
 };
 
+#ifdef HAVE_LIBDECOR_H
+static void
+decoration_frame_configure(struct libdecor_frame *frame,
+                           struct libdecor_configuration *configuration,
+                           void *user_data)
+{
+    SDL_WindowData *wind = user_data;
+    SDL_Window *window = wind->sdlwindow;
+    int width, height;
+    enum libdecor_window_state window_state;
+    struct libdecor_state *state;
+
+    /* window size */
+    if (!libdecor_configuration_get_content_size(configuration, frame, &width, &height)) {
+        width = window->w;
+        height = window->h;
+    }
+
+    wind->resize.width = width;
+    wind->resize.height = height;
+
+    wind->resize.pending = SDL_TRUE;
+    wind->resize.configure = SDL_TRUE;
+    Wayland_HandlePendingResize(window);
+    wind->shell_surface.libdecor.initial_configure_seen = SDL_TRUE;
+
+    window->w = wind->resize.width;
+    window->h = wind->resize.height;
+
+    /* window state */
+    if (!libdecor_configuration_get_window_state(configuration, &window_state)) {
+        window_state = LIBDECOR_WINDOW_STATE_NONE;
+    }
+
+    if (window_state & LIBDECOR_WINDOW_STATE_MAXIMIZED) {
+        window->flags |= SDL_WINDOW_MAXIMIZED;
+        SDL_SendWindowEvent(window, SDL_WINDOWEVENT_MAXIMIZED, 0, 0);
+    }
+    else {
+        window->flags &= ~SDL_WINDOW_MAXIMIZED;
+        SDL_SendWindowEvent(window, SDL_WINDOWEVENT_RESTORED, 0, 0);
+    }
+
+    if (window_state & LIBDECOR_WINDOW_STATE_ACTIVE) {
+        window->flags |= SDL_WINDOW_INPUT_FOCUS;
+        SDL_SendWindowEvent(window, SDL_WINDOWEVENT_FOCUS_GAINED, 0, 0);
+    }
+    else {
+        window->flags &= ~SDL_WINDOW_INPUT_FOCUS;
+        SDL_SendWindowEvent(window, SDL_WINDOWEVENT_FOCUS_LOST, 0, 0);
+    }
+
+    /* The fullscreen flag is already set my some other entity when 'SDL_SetWindowFullscreen'
+     * is called, but we will set it here again in case the compositor requests fullscreen.
+     */
+    if (window_state & LIBDECOR_WINDOW_STATE_FULLSCREEN) {
+        window->flags |= SDL_WINDOW_FULLSCREEN;
+    }
+    else {
+        window->flags &= ~SDL_WINDOW_FULLSCREEN;
+    }
```

Comment:
Redid this to match xdg-shell and fix bugs related to redundancy, which caused events to not fire.

```suggestion
   /* Always send maximized/restored/focus events; if the event is redundant it will
     * automatically be discarded (see src/events/SDL_windowevents.c).
     *
     * No, we do not get minimize events from libdecor.
     */
    if (!(window_state & LIBDECOR_WINDOW_STATE_FULLSCREEN)) {
        window->flags &= ~SDL_WINDOW_FULLSCREEN_DESKTOP;
        SDL_SendWindowEvent(window,
                            (window_state & LIBDECOR_WINDOW_STATE_MAXIMIZED) ?
                                SDL_WINDOWEVENT_MAXIMIZED :
                                SDL_WINDOWEVENT_RESTORED,
                            0, 0);
    } else {
        window->flags |= SDL_WINDOW_FULLSCREEN_DESKTOP;
    }
    SDL_SendWindowEvent(window,
                        (window_state & LIBDECOR_WINDOW_STATE_ACTIVE) ?
                            SDL_WINDOWEVENT_FOCUS_GAINED :
                            SDL_WINDOWEVENT_FOCUS_LOST,
                        0, 0);
```